rclpy/rclpy/node.py

```diff
@@ -217,12 +230,82 @@ def get_logger(self):
         """Get the nodes logger."""
         return self._logger
 
+    def declare_parameter(self, parameter: Parameter) -> Parameter:
+        """
+        Declare and initialize a parameter.
+
+        This method, if successful, will result in any callback registered with
+        :func:`set_parameters_callback` to be called.
+
+        :param parameter: The parameter to declare, with a name, value and descriptor.
+        :return: Parameter with the effectively assigned value.
+        :raises: ParameterAlreadyDeclaredException if the parameter had already been declared.
+        :raises: InvalidParameterException if the parameter name is invalid.
+        :raises: InvalidParameterValueException if the registered callback rejects the parameter.
+        """
+        return self.declare_parameters('/', [parameter])
+
+    def declare_parameters(self, namespace: str, parameters: List[Parameter]) -> List[Parameter]:
+        """
+        Declare a list of parameters.
+
+        This method, if successful, will result in any callback registered with
+        :func:`set_parameters_callback` to be called once for each parameter.
+        If one of those calls fail, an exception will be raised and the remaining parameters will
+        not be declared. Parameters declared up to that point will not be undeclared.
+
+        :param namespace: Namespace for parameters.
+        :param parameters: The parameters to declare, with a name, value and descriptor.
+        :return: Parameter list with the effectively assigned values for each of them.
+        :raises: ParameterAlreadyDeclaredException if the parameter had already been declared.
+        :raises: InvalidParameterException if the parameter name is invalid.
+        :raises: InvalidParameterValueException if the registered callback rejects the parameter.
+        """
+        for parameter in parameters:
+            parameter.name = namespace + parameter.name
```

Comment:
Writing tests I just realized that this is not possible (name can't be set after creating `Parameter` object).
I can think of these alternatives, in order of preference:
- Creating new parameters with same value, type and descriptor but with a new name containing the namespace.
- Suppressing `namespace` parameter. This is introduced to match the behavior in `rclcpp`, but on the other hand the current `set_parameters` method in `master` doesn't have a namespace option.
- Adding a setter to the name.
- Using the private `_name` member (argh).

@hidmic thoughts?
/cc @wjwwood input on this is welcome!

Reply:
Why not have a version of `set_parameters` or `self._set_parameters` which takes a tuple of `(name, value)`, then you can take apart the `Parameter` and use it how you want.

I guess that's the same as creating a temporary `Parameter` with the name you want...

Reply:
Considering that other parameter-related methods typically reference parameters by name (undeclare, get, set), it might make sense to also accept names when declaring parameters as well. Like @wjwwood suggests, we could change `set_parameters` to also take a tuple of `(name, value)`. This would be more akin to the implementation in `rclcpp`.

We use a temporary parameter `Parameter` and create a new one with the provided name, e.g

```python
def declare_parameter(self, name: str, parameter: Parameter) -> Parameter:
    ...
def declare_parameters(self, namespace: str, parameters: List[Tuple[str, Parameter]]) -> List[Parameter]:
    ...
```

Or consider using a `ParameterValue` instead (like `rclcpp`):

```python
def declare_parameter(self, name: str, parameter: ParameterValue) -> Parameter:
    ...
def declare_parameters(self, namespace: str, parameters: List[Tuple[str, ParameterValue]]) -> List[Parameter]:
    ...
```

Reply:
See a proposed fix in https://github.com/ros2/rclpy/pull/325/commits/a39ea62cb69685e437a5935cfc206a648808d26a.

In `rclcpp`, `declare_parameter` takes a name, a value and a descriptor, but `set_parameter` takes a `Parameter_T`. Following that line, I left `set_parameters` with the same signature, and changed `declare_parameters` to take the pieces of the parameter as separate arguments in a tuple.
In other words, its @jacobperron 's second suggestion, considering `ParameterDescriptor` as well.
```
def declare_parameter(self, name: str, parameter: ParameterValue, descriptor: ParameterDescriptor) -> Parameter:
    ...
def declare_parameters(self, namespace: str, parameters: List[Tuple[str, ParameterValue, ParameterDescriptor]]) -> List[Parameter]:
    ...
```
/cc @hidmic

Reply:
For what it is worth, there's no reason C++ doesn't provide a `set_parameters(vector<pair<string, ParameterValue>>)` nor is there a reason against `declare_parameters(string, vector<pair<Parameter, descriptor>>)`, other than that I ran out of time to keep adding convenient signatures. Many of the other signatures that seem less obvious were added by request before my big refactor.